Infrastructure: replace `qsl(...)` used for Tooltips

SlySven opened the pull request:
Specifically when it is used to put HTML like tags around the text so that it triggers the "rich-text" detector in the Qt libraries so that the tips gets better formatting than being a single (long) line of plain text.

Whilst working on this PR I also spotted and fixed:
* A couple of raw strings in the `VarUnit` class that should have been translated (or replaced with a `QString()`) - this required adding the `Q_DECLARE_TR_FUNCTIONS(`className`)` macro to that class as it is not derived from the `QObject` class which we normally use to provide the `QObject::tr(`...`)` method.
* ~~The `XMLImport` class had an extrenuous ';' in the header file where it had a `Q_DECLARE_TR_FUNCTIONS(XMLImport)` macro.~~
* The tooltip for the "Save item" button in the editor contained a unmatched `<p>`...`</p>` tag pair as well as a useless (as it is ignored there) `\n` {Line Feed}.
* The tooltip for the "Save profile" button in the editor contained a unmatched `<p>`...`</p>` tag pair.
* I had previously defined the `QT_NO_CAST_FROM_ASCII` and `QT_NO_CAST_TO_ASCII` macros in the `dlgRoomExits` class ~~but since the `mudlet.h` needs to be included for that class to give access to the `mudlet::htmlWrapper(`...`)` that this PR now uses more widely that was triggering warnings about raw string literals in *I think* third party library code.~~ *but they are no longer considered useful for us* ~~S~~so I ~~had to remove~~*have removed* those Qt macros.
* ~~Generally, tool-tips are sentences so they should end with terminal punctuation, usually a '.' and there were quite a few without those, they have been fixed!~~
* The ~~`mudlet::htmlWrapper`~~*`utils::richText`*`(`...`)` has exactly the same code effect as the `(const QString) singleParagraph` helper function in the `dlgRoomExits` class - so that has been replaced by it.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

#### Release post highlight
Infrastructure so probably none...!

***Edited: to reflect content that has been reverted from this PR.***

SlySven committed:
Infrastructure (Public): replace QStringLiteral used for Tooltips

Specifically when it is used to put HTML like tags around the text so that it
triggers the "rich-text" detector in the Qt libraries so that the tips gets
better formatting than being a single (long) line of plain text.

Whilst working on this PR I also spotted and fixed:
* A couple of raw strings in the `VarUnit` class that should have been
translated (or replaced with a `QString()`) - this required adding the
`Q_DECLARE_TR_FUNCTIONS(`className`)` macro to that class as it is not
derived from the `QObject` class which we normally use to provide the
`QObject::tr(`...`)` method.
* The `XMLImport` class had an extrenuous ';' in the header file where
it had a `Q_DECLARE_TR_FUNCTIONS(XMLImport)` macro.
* The tooltip for the "Save item" button in the editor contained a
unmatched `<p>`...`</p>` tag pair as well as a useless (as it is ignored
there) `\n` {Line Feed}.
* The tooltip for the "Save profile" button in the editor contained a
unmatched `<p>`...`</p>` tag pair.
* I had previously defined the `QT_NO_CAST_FROM_ASCII` and
`QT_NO_CAST_TO_ASCII` macros in the `dlgRoomExits` class but since the
`mudlet.h` needs to be included for that class to give access to the
`mudlet::htmlWrapper(`...`)` that this PR now uses more widely that was
triggering warnings about raw string literals in *I think* third party
library code. So I had to remove those Qt macros.
* Generally, tool-tips are sentences so they should end with terminal
punctuation, usually a '.' and there were quite a few without those, they
have been fixed!
* The `mudlet::htmlWrapper(`...`)` has exactly the same code effect as
the `(const QString) singleParagraph` helper function in the `dlgRoomExits`
class - so it has been replaced by that.

This should close Issue #5691 but will cause #3952 to need tweaking to
at least remove a TODO that mentions that Issue!

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

demonnic commented:
> Whilst working on this PR I also spotted and fixed:

I worry when I see this in a PR. Especially one that touches 15 files. 

I feel like adding terminal characters to tooltips is its own PR.
Replacing an entire function in another class with this new one should probably be a separate PR
Possibly also refactoring items such as removing an unneeded (but not code breaking) `;` 

As it stands this seems like a daunting amount of diff to look through for errors in review, once it passes compiling in CI.

SlySven committed:
Fix: correct typos I'd left in untested code

These were in parts associated with the Updater code which I normally
have disabled in my local builds!

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Partial revert: remove some non-core changes from PR

These are still desirable in my opinion but peers reviewing the changes
wanted them to be moved to separate PR(s).

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
> I feel like adding terminal characters to tooltips is its own PR.

Well in those cases the *terminal character* i.e. `'.'` goes in the **text** on the same line as is being edited by adding the `mudlet::htmlWrapper(`...`)` that this PR is centred around - this means that a *separate* PR **cannot** be done until this PR is merged as they will act on the same line and thus conflict - and also (AFAICR) they will not be adding any more lines to the change-set.

> Replacing an entire function in another class with this new one should probably be a separate PR

Not really as the (one-line) function in this PR does across many classes that which that (one-line) helper function did in only one.

> Possibly also refactoring items such as removing an unneeded (but not code breaking) ;

Maybe, but I only spotted that it was there (and redundant) when I was checking on the usage of the Qt Macro across all our class - and for a single byte it seemed a bit OTT to do a whole PR for.

vadi2 commented:
Things like periods at the end of tooltips are a personal preference that has nothing to do with a purely technical refactoring, so that will generate discussion completely unrelated to the task. I too have PRs that are waiting to be merged before I can continue working more on them, that's not a reason to stuff everything into one.

We agreed that refactoring PRs should do one thing and one thing only. https://github.com/Mudlet/Mudlet/pull/5694 is a good example of this - it is very easy to scan through the PR to verify it does what it should and test it. I agree with demonnic, I'm not up for dissecting these changes in this format - please redo these separately and it'll be much easier to push them through.

SlySven committed:
Fix: typos in a help/guidance file

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Fix: correct a typo causing a failure to build

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Fix: a bit of barrel scraping - restore an #include

I took it out because I believed it was included anyway by another header
`./src/mudlet.h` that needed to go in.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Fix: back out a FTB issue

Will need to file an issue so as to shut the Danger Bot up about the TODO
I've inserted to do the change I had to leave incomplete!

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Fix: back out even more of a change that is breaking CMake CI builds

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into Infrastructure(Public)_replace_QStringLiteral_usedForTooltips

In case *this* was why the CI builds were failing...

Signed-off by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Finally: have spotted problem and fixed it

An `#include` of `Host.h` had slipped into `./src/VarUnit.cpp`!

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
@demonnic of the three points listed in https://github.com/Mudlet/Mudlet/pull/5694#issuecomment-979416226 I had done the first and the third - but left the second one as it was as I think it is still valid as part of ***this*** PR. Would you like to take a gander at it now?

demonnic commented on src/VarUnit.h:
```diff
@@ -23,6 +24,7 @@
 
 
 #include "pre_guard.h"
+#include <QApplication>
```
I'm not entirely sure why this include is added?

SlySven replied:
For the `Q_DECLARE_TR_FUNCTIONS(` ... classname ... `)` macro which is needed to get the `tr(`...`)` method/function to put strings through the Qt translation system if a class is NOT derived from the `QObject` class (in ***those** cases you'd use the `QOBJECT` macro) - this is a (lighter, I guess) macro that only adds in stuff for the translation system without all the other things (Signal/Slots, metaproperties) that QOBJECT provides.

The bottom line is that I found a single UI string in this class - and that wasn't being translated.

demonnic replied:
Good enough for me, thanks for explaining =)

demonnic reviewed (commented):
I think there was only the one thing which really stood out. Apart from that, it seems to be functional, I didn't notice any particularly broken tooltips while browsing them in the test version. 

I'm still not sure that I agree that "short pieces of text that can give additional hints or help" automatically means that text is a sentence and should automatically be terminated as such, but I don't feel strongly enough about it to make it a blocker

vadi2 commented:
Periods have no places in tooltips, it seems. Something about prioritising brevity 🤭 https://ux.stackexchange.com/a/141723/108628

SlySven commented:
> Periods have no places in tooltips, it seems...

That is not the whole answer that was given was it! In summary it says:

![image](https://i.stack.imgur.com/tajxW.png)

vadi2 commented:
Exactly

SlySven commented:
~~Anyhow this one will have to wait until the fate of #5640 is resolved...~~ *That has now happened and I have incorporated the effects of it in this one!*

SlySven committed:
Merge branch 'development' into Infrastructure(Public)_replace_QStringLiteral_usedForTooltips

Conflicts resolved in:
* src/TConsole.cpp
* src/TMainConsole.cpp
* src/TTextEdit.cpp
* src/dlgConnectionProfiles.cpp
* src/dlgModuleManager.cpp
* src/dlgProfilePreferences.cpp
* src/dlgRoomExits.cpp
* src/dlgTriggerEditor.cpp
* src/mudlet.cpp
* src/mudlet.h

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: move/rename mudlet::htmlWrapper to utils::richText

This requires making the `./src/utils.h` into a trivial class but allows
the function/method to also be used within the `VarUnit` class...

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Update CONTRIBUTING.md

Update text now that the function is called `utils::richText` rather than
the original (and with slightly more characters in the identifier)
`mudlet::htmlWrapper`.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Update dlgRoomExits.cpp

Revert header file `#include` that is no longer needed - but insert a
replacement one that is, instead.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Update T2DMap.cpp

Revert a tooltip change that had not previously been backed out.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 commented:
> The general principle is to mercilessly cut down the amount of words / characters etc, prioritizing brevity.

I thought we just agreed that tooltips - the single sentence ones at least - should not have periods in them? Of course multiple sentences should have them, but then again tooltips should not be multiple-sentence long paragraphs!

SlySven commented:
> I thought we just agreed that tooltips - the single sentence ones at least - should not have periods in them? Of course multiple sentences should have them, but then again tooltips should not be multiple-sentence long paragraphs!

I have cut back on the changes in this PR so the tool-tip texts (try saying THAT fast!) should be unchanged apart from the removal or addition of HTML tags as appropriate.

Further editing of the `CONTRIBUTING.md` file once we have nailed down precisely what it should be is, I guess, a possibility, but this PR is primarily hoping to make the generation of tool-tips more consistent (and use less typing) - and for single paragraphs (or less), remove the need for translators to have to deal with extraneous (for them) `<p>`...`</p>` tags. 🤞

vadi2 committed:
Merge branch 'development' into Infrastructure(Public)_replace_QStringLiteral_usedForTooltips
